In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('/content/Housing.csv')

In [3]:
df

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished
...,...,...,...,...,...,...,...,...,...,...,...,...,...
540,1820000,3000,2,1,1,yes,no,yes,no,no,2,no,unfurnished
541,1767150,2400,3,1,1,no,no,no,no,no,0,no,semi-furnished
542,1750000,3620,2,1,1,yes,no,no,no,no,0,no,unfurnished
543,1750000,2910,3,1,1,no,no,no,no,no,0,no,furnished


In [35]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler , OneHotEncoder , LabelEncoder

In [36]:
le = LabelEncoder()
features = ['mainroad',	'guestroom',	'basement',	'hotwaterheating'	,'airconditioning', 'prefarea','furnishingstatus']

for i in features:
  df[i] = le.fit_transform(df[i])

In [52]:
df

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,1,0,0,0,1,2,1,0
1,12250000,8960,4,4,4,1,0,0,0,1,3,0,0
2,12250000,9960,3,2,2,1,0,1,0,0,2,1,1
3,12215000,7500,4,2,2,1,0,1,0,1,3,1,0
4,11410000,7420,4,1,2,1,1,1,0,1,2,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
540,1820000,3000,2,1,1,1,0,1,0,0,2,0,2
541,1767150,2400,3,1,1,0,0,0,0,0,0,0,1
542,1750000,3620,2,1,1,1,0,0,0,0,0,0,2
543,1750000,2910,3,1,1,0,0,0,0,0,0,0,0


In [37]:
X = df.drop('price',axis=1)
y = df['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [38]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score , root_mean_squared_error

In [53]:
lr = LinearRegression()
lr.fit(X_train,y_train)
y_pred = lr.predict(X_test)

print(r2_score(y_test,y_pred))
print(mean_squared_error(y_test,y_pred))

0.6494754192267803
1771751116594.0352


In [54]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
    ('encoder',LabelEncoder(),['mainroad',	'guestroom',	'basement',	'hotwaterheating'	,'airconditioning', 'prefarea','furnishingstatus']),
    ('scaler',StandardScaler(),['area']),
    ('model',LinearRegression())
],verbose=True)

In [72]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Define a list of models to evaluate
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree Regressor": DecisionTreeRegressor(random_state=42),
    "Random Forest Regressor": RandomForestRegressor(random_state=42)
}

results = {}

for name, model in models.items():
    print(f"\n--- Evaluating {name} ---")
    # Create a pipeline with StandardScaler and the current model
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('model', model)
    ])

    # Fit the pipeline
    pipeline.fit(X_train, y_train)

    # Make predictions
    y_pred = pipeline.predict(X_test)

    # Evaluate the model
    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)

    print(f"R-squared: {r2}")
    print(f"Mean Squared Error: {mse}")

    results[name] = {"R-squared": r2, "Mean Squared Error": mse}

# Optionally, find and print the best model based on R-squared
best_model_name = max(results, key=lambda k: results[k]['R-squared'])
print(f"\n--- Best Model based on R-squared: {best_model_name} ---")
print(f"R-squared: {results[best_model_name]['R-squared']}")
print(f"Mean Squared Error: {results[best_model_name]['Mean Squared Error']}")


--- Evaluating Linear Regression ---
R-squared: 0.6494754192267792
Mean Squared Error: 1771751116594.041

--- Evaluating Decision Tree Regressor ---
R-squared: 0.4681689010023007
Mean Squared Error: 2688177649082.569

--- Evaluating Random Forest Regressor ---
R-squared: 0.6118077610803518
Mean Squared Error: 1962144940711.022

--- Best Model based on R-squared: Linear Regression ---
R-squared: 0.6494754192267792
Mean Squared Error: 1771751116594.041


In [73]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Define a list of models to evaluate, including a new ensemble technique
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree Regressor": DecisionTreeRegressor(random_state=42),
    "Random Forest Regressor": RandomForestRegressor(random_state=42),
    "Gradient Boosting Regressor": GradientBoostingRegressor(random_state=42)
}

results = {}

print("--- Evaluating Models with StandardScaler ---")
for name, model in models.items():
    print(f"\n--- Evaluating {name} ---")
    # Create a pipeline with StandardScaler and the current model
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('model', model)
    ])

    # Fit the pipeline
    pipeline.fit(X_train, y_train)

    # Make predictions
    y_pred = pipeline.predict(X_test)

    # Evaluate the model
    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)

    print(f"R-squared: {r2}")
    print(f"Mean Squared Error: {mse}")

    results[name] = {"R-squared": r2, "Mean Squared Error": mse}

# Find and print the best model based on R-squared
best_model_name = max(results, key=lambda k: results[k]['R-squared'])
print(f"\n--- Best Model based on R-squared: {best_model_name} ---")
print(f"R-squared: {results[best_model_name]['R-squared']}")
print(f"Mean Squared Error: {results[best_model_name]['Mean Squared Error']}")

--- Evaluating Models with StandardScaler ---

--- Evaluating Linear Regression ---
R-squared: 0.6494754192267792
Mean Squared Error: 1771751116594.041

--- Evaluating Decision Tree Regressor ---
R-squared: 0.4681689010023007
Mean Squared Error: 2688177649082.569

--- Evaluating Random Forest Regressor ---
R-squared: 0.6118077610803518
Mean Squared Error: 1962144940711.022

--- Evaluating Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102

--- Best Model based on R-squared: Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102


### Evaluating XGBoost Regressor

Now, let's include the `XGBoost Regressor` in our comparison to see if it can further improve performance.

In [75]:
from xgboost import XGBRegressor

# Add XGBoost Regressor to the models dictionary
models["XGBoost Regressor"] = XGBRegressor(random_state=42)

print("--- Evaluating Models with StandardScaler (including XGBoost) ---")
for name, model in models.items():
    # Skip already evaluated models to avoid redundant computations, unless it's a new run
    if name not in results or (name == "XGBoost Regressor" and "XGBoost Regressor" not in results):
        print(f"\n--- Evaluating {name} ---")
        # Create a pipeline with StandardScaler and the current model
        pipeline = Pipeline([
            ('scaler', StandardScaler()),
            ('model', model)
        ])

        # Fit the pipeline
        pipeline.fit(X_train, y_train)

        # Make predictions
        y_pred = pipeline.predict(X_test)

        # Evaluate the model
        r2 = r2_score(y_test, y_pred)
        mse = mean_squared_error(y_test, y_pred)

        print(f"R-squared: {r2}")
        print(f"Mean Squared Error: {mse}")

        results[name] = {"R-squared": r2, "Mean Squared Error": mse}

# Find and print the best model based on R-squared
best_model_name = max(results, key=lambda k: results[k]['R-squared'])
print(f"\n--- Best Model based on R-squared: {best_model_name} ---")
print(f"R-squared: {results[best_model_name]['R-squared']}")
print(f"Mean Squared Error: {results[best_model_name]['Mean Squared Error']}")

--- Evaluating Models with StandardScaler (including XGBoost) ---

--- Evaluating XGBoost Regressor ---
R-squared: 0.597907543182373
Mean Squared Error: 2032404398080.0

--- Best Model based on R-squared: Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102


### Bagging and Boosting Techniques

`Gradient Boosting Regressor` and `XGBoost Regressor` are both powerful boosting techniques that build an ensemble of weak learners sequentially, with each new learner correcting errors made by the previous ones.

**Bagging (Bootstrap Aggregating)**, on the other hand, involves training multiple models (usually of the same type) independently on different bootstrap samples of the training data, and then combining their predictions (e.g., by averaging for regression). A common example is the `BaggingRegressor` from scikit-learn.

While Gradient Boosting and XGBoost are highly effective, explicitly trying a `BaggingRegressor` with a strong base estimator (like a `DecisionTreeRegressor`) can sometimes provide different insights or performance, especially if the individual models are diverse.

Let's evaluate a `BaggingRegressor`.

In [76]:
from sklearn.ensemble import BaggingRegressor

# Add Bagging Regressor to the models dictionary
# Using a Decision Tree Regressor as the base estimator
models["Bagging Regressor (Decision Tree)"] = BaggingRegressor(
    estimator=DecisionTreeRegressor(random_state=42),
    n_estimators=100, # Number of base estimators
    random_state=42,
    n_jobs=-1 # Use all available cores
)

print("--- Evaluating Models with StandardScaler (including Bagging) ---")
# Re-evaluating all models to ensure 'results' dictionary is updated with Bagging Regressor
for name, model in models.items():
    print(f"\n--- Evaluating {name} ---")
    # Create a pipeline with StandardScaler and the current model
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('model', model)
    ])

    # Fit the pipeline
    pipeline.fit(X_train, y_train)

    # Make predictions
    y_pred = pipeline.predict(X_test)

    # Evaluate the model
    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)

    print(f"R-squared: {r2}")
    print(f"Mean Squared Error: {mse}")

    results[name] = {"R-squared": r2, "Mean Squared Error": mse}

# Find and print the best model based on R-squared
best_model_name = max(results, key=lambda k: results[k]['R-squared'])
print(f"\n--- Best Model based on R-squared: {best_model_name} ---")
print(f"R-squared: {results[best_model_name]['R-squared']}")
print(f"Mean Squared Error: {results[best_model_name]['Mean Squared Error']}")

--- Evaluating Models with StandardScaler (including Bagging) ---

--- Evaluating Linear Regression ---
R-squared: 0.6494754192267792
Mean Squared Error: 1771751116594.041

--- Evaluating Decision Tree Regressor ---
R-squared: 0.4681689010023007
Mean Squared Error: 2688177649082.569

--- Evaluating Random Forest Regressor ---
R-squared: 0.6118077610803518
Mean Squared Error: 1962144940711.022

--- Evaluating Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102

--- Evaluating XGBoost Regressor ---
R-squared: 0.597907543182373
Mean Squared Error: 2032404398080.0

--- Evaluating Bagging Regressor (Decision Tree) ---
R-squared: 0.6089273838500855
Mean Squared Error: 1976704009757.4185

--- Best Model based on R-squared: Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102


### Final Model Pipeline for Prediction

Based on our evaluation, the **Gradient Boosting Regressor** (untuned) demonstrated the best performance with an R-squared of approximately 0.665. Let's create a final, clean pipeline using this model, incorporating the `StandardScaler` that was used during training.

This pipeline can then be used to make predictions on new, unseen data.

In [77]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

# Instantiate the best model
best_model = GradientBoostingRegressor(random_state=42)

# Create the final pipeline
final_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('model', best_model)
])

# Train the final pipeline on the full training data
final_pipeline.fit(X_train, y_train)

print("Final Gradient Boosting Regressor pipeline trained successfully.")

Final Gradient Boosting Regressor pipeline trained successfully.


### Making Predictions on New Data

Now, let's simulate making predictions on some new, unseen data. We'll use a small sample from our `X_test` set for demonstration purposes.

In [79]:
import numpy as np

# Take a small sample from the test set to simulate new data
# We'll take the first 5 rows of X_test as 'new_data'
new_data_sample = X_test.head(5)

print("New data sample for prediction:")
display(new_data_sample)

# Make predictions using the final pipeline
predictions = final_pipeline.predict(new_data_sample)

print("\nPredictions for the new data sample:")
for i, pred in enumerate(predictions):
    print(f"Sample {i+1}: Predicted Price = {pred:,.2f}")

# You can compare these to the actual prices if you have them for this sample
actual_prices_sample = y_test.head(5)
print("\nActual prices for the new data sample:")
display(actual_prices_sample)

New data sample for prediction:


,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
316,5900,4,2,2,0,0,1,0,0,1,0,2
77,6500,3,2,3,1,0,0,0,1,0,1,0
360,4040,2,1,1,1,0,0,0,0,0,0,1
90,5000,3,1,2,1,0,0,0,1,0,0,1
493,3960,3,1,1,1,0,0,0,0,0,0,0



Predictions for the new data sample:
Sample 1: Predicted Price = 4,549,951.16
Sample 2: Predicted Price = 7,319,875.02
Sample 3: Predicted Price = 3,709,740.35
Sample 4: Predicted Price = 4,364,303.62
Sample 5: Predicted Price = 3,720,826.43

Actual prices for the new data sample:


,price
316,4060000
77,6650000
360,3710000
90,6440000
493,2800000


### Applying Bagging with Ridge and Lasso Regressors

Let's expand our ensemble exploration by applying `BaggingRegressor` with `Ridge` and `Lasso` as its base estimators. These are regularized linear models that can help prevent overfitting, and bagging them might yield more robust predictions.

In [81]:
from sklearn.linear_model import Ridge, Lasso
from sklearn.ensemble import BaggingRegressor

# Add Bagging Regressor with Ridge as base estimator
models["Bagging Regressor (Ridge)"] = BaggingRegressor(
    estimator=Ridge(random_state=42),
    n_estimators=100, # Number of base estimators
    random_state=42,
    n_jobs=-1 # Use all available cores
)

# Add Bagging Regressor with Lasso as base estimator
models["Bagging Regressor (Lasso)"] = BaggingRegressor(
    estimator=Lasso(random_state=42, max_iter=2000), # Increased max_iter for convergence
    n_estimators=100, # Number of base estimators
    random_state=42,
    n_jobs=-1 # Use all available cores
)

print("--- Evaluating Models with StandardScaler (including Bagging with Ridge/Lasso) ---")
# Re-evaluating all models to ensure 'results' dictionary is updated with new Bagging Regressors
for name, model in models.items():
    print(f"\n--- Evaluating {name} ---")
    # Create a pipeline with StandardScaler and the current model
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('model', model)
    ])

    # Fit the pipeline
    pipeline.fit(X_train, y_train)

    # Make predictions
    y_pred = pipeline.predict(X_test)

    # Evaluate the model
    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)

    print(f"R-squared: {r2}")
    print(f"Mean Squared Error: {mse}")

    results[name] = {"R-squared": r2, "Mean Squared Error": mse}

# Find and print the best model based on R-squared
best_model_name = max(results, key=lambda k: results[k]['R-squared'])
print(f"\n--- Best Model based on R-squared: {best_model_name} ---")
print(f"R-squared: {results[best_model_name]['R-squared']}")
print(f"Mean Squared Error: {results[best_model_name]['Mean Squared Error']}")

--- Evaluating Models with StandardScaler (including Bagging with Ridge/Lasso) ---

--- Evaluating Linear Regression ---
R-squared: 0.6494754192267792
Mean Squared Error: 1771751116594.041

--- Evaluating Decision Tree Regressor ---
R-squared: 0.4681689010023007
Mean Squared Error: 2688177649082.569

--- Evaluating Random Forest Regressor ---
R-squared: 0.6118077610803518
Mean Squared Error: 1962144940711.022

--- Evaluating Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102

--- Evaluating XGBoost Regressor ---
R-squared: 0.597907543182373
Mean Squared Error: 2032404398080.0

--- Evaluating Bagging Regressor (Decision Tree) ---
R-squared: 0.6089273838500855
Mean Squared Error: 1976704009757.4185

--- Evaluating Bagging Regressor (Ridge) ---
R-squared: 0.6495704690016202
Mean Squared Error: 1771270680830.2065

--- Evaluating Bagging Regressor (Lasso) ---
R-squared: 0.6496870812891449
Mean Squared Error: 1770681255831.3245

--- Best Mode

### Applying Stacking Ensemble with Gradient Boosting Regressor as Meta-Model

Stacking is a powerful ensemble technique where the predictions of several base estimators (level-0 models) are used as input features for a meta-estimator (level-1 model). We'll use `LinearRegression`, `RandomForestRegressor`, and `XGBRegressor` as our base models, and a `GradientBoostingRegressor` as the final meta-model to make the ultimate prediction.

In [82]:
from sklearn.ensemble import StackingRegressor, GradientBoostingRegressor, RandomForestRegressor
from sklearn.linear_model import LinearRegression
from xgboost import XGBRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import r2_score, mean_squared_error

# Define base estimators (level-0 models), each with its own StandardScaler
# This ensures that each base model receives scaled data independently
estimators = [
    ('lr', Pipeline([
        ('scaler', StandardScaler()),
        ('model', LinearRegression())
    ])),
    ('rf', Pipeline([
        ('scaler', StandardScaler()),
        ('model', RandomForestRegressor(random_state=42))
    ])),
    ('xgb', Pipeline([
        ('scaler', StandardScaler()),
        ('model', XGBRegressor(random_state=42))
    ]))
]

# Define the meta-estimator (level-1 model)
# The GradientBoostingRegressor will learn to combine the predictions of the base estimators
final_estimator = GradientBoostingRegressor(random_state=42)

# Create the StackingRegressor pipeline
# This pipeline will handle the training of base estimators and the meta-estimator
stacked_model = StackingRegressor(
    estimators=estimators,
    final_estimator=final_estimator,
    cv=5, # Number of cross-validation folds to train base models
    n_jobs=-1 # Use all available cores
)

# Add the stacked model to our dictionary for evaluation
models["Stacking Regressor (GBM Meta)"] = stacked_model

print("--- Evaluating Models (including Stacking Regressor) ---")
# Re-evaluating all models to ensure 'results' dictionary is updated with the new stacked model
for name, model in models.items():
    print(f"\n--- Evaluating {name} ---")
    # For stacking, the internal pipelines for base estimators already handle scaling.
    # So, we'll create a simple pipeline here that just contains the model if it's already a pipeline (like StackingRegressor).
    # For other models, we'll keep the StandardScaler pipeline as before.
    if isinstance(model, StackingRegressor):
        pipeline = model # The stacking regressor itself is a complex pipeline
    else:
        pipeline = Pipeline([
            ('scaler', StandardScaler()),
            ('model', model)
        ])

    # Fit the pipeline
    pipeline.fit(X_train, y_train)

    # Make predictions
    y_pred = pipeline.predict(X_test)

    # Evaluate the model
    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)

    print(f"R-squared: {r2}")
    print(f"Mean Squared Error: {mse}")

    results[name] = {"R-squared": r2, "Mean Squared Error": mse}

# Find and print the best model based on R-squared
best_model_name = max(results, key=lambda k: results[k]['R-squared'])
print(f"\n--- Best Model based on R-squared: {best_model_name} ---")
print(f"R-squared: {results[best_model_name]['R-squared']}")
print(f"Mean Squared Error: {results[best_model_name]['Mean Squared Error']}")

--- Evaluating Models (including Stacking Regressor) ---

--- Evaluating Linear Regression ---
R-squared: 0.6494754192267792
Mean Squared Error: 1771751116594.041

--- Evaluating Decision Tree Regressor ---
R-squared: 0.4681689010023007
Mean Squared Error: 2688177649082.569

--- Evaluating Random Forest Regressor ---
R-squared: 0.6118077610803518
Mean Squared Error: 1962144940711.022

--- Evaluating Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102

--- Evaluating XGBoost Regressor ---
R-squared: 0.597907543182373
Mean Squared Error: 2032404398080.0

--- Evaluating Bagging Regressor (Decision Tree) ---
R-squared: 0.6089273838500855
Mean Squared Error: 1976704009757.4185

--- Evaluating Bagging Regressor (Ridge) ---
R-squared: 0.6495704690016202
Mean Squared Error: 1771270680830.2065

--- Evaluating Bagging Regressor (Lasso) ---
R-squared: 0.6496870812891449
Mean Squared Error: 1770681255831.3245

--- Evaluating Stacking Regressor (GBM 

### Advanced Hyperparameter Tuning with Optuna for GradientBoostingRegressor

To further optimize the `GradientBoostingRegressor`, we'll use Optuna, an automatic hyperparameter optimization framework. Optuna intelligently searches for the best combination of hyperparameters to minimize (or maximize) a given objective function. In our case, the objective will be to minimize the Mean Squared Error (MSE) on the validation set.

In [83]:
# Install Optuna if not already installed
!pip install optuna --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 419.5/419.5 kB 8.3 MB/s eta 0:00:00


In [85]:
import optuna
from sklearn.model_selection import KFold
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

def objective(trial):
    # Define the search space for GradientBoostingRegressor hyperparameters
    n_estimators = trial.suggest_int('n_estimators', 100, 1000, step=100)
    learning_rate = trial.suggest_float('learning_rate', 0.01, 0.2, log=True)
    max_depth = trial.suggest_int('max_depth', 3, 10)
    min_samples_split = trial.suggest_int('min_samples_split', 2, 20)
    min_samples_leaf = trial.suggest_int('min_samples_leaf', 1, 10)
    subsample = trial.suggest_float('subsample', 0.6, 1.0, step=0.1)

    # Create a GradientBoostingRegressor with the suggested hyperparameters
    gbm = GradientBoostingRegressor(
        n_estimators=n_estimators,
        learning_rate=learning_rate,
        max_depth=max_depth,
        min_samples_split=min_samples_split,
        min_samples_leaf=min_samples_leaf,
        subsample=subsample,
        random_state=42
    )

    # Create a pipeline with StandardScaler and the GradientBoostingRegressor
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('model', gbm)
    ])

    # Use K-Fold cross-validation to evaluate the model
    kf = KFold(n_splits=5, shuffle=True, random_state=42)
    mse_scores = []

    for train_index, val_index in kf.split(X_train):
        X_train_fold, X_val_fold = X_train.iloc[train_index], X_train.iloc[val_index]
        y_train_fold, y_val_fold = y_train.iloc[train_index], y_train.iloc[val_index]

        pipeline.fit(X_train_fold, y_train_fold)
        y_pred_val = pipeline.predict(X_val_fold)
        mse_scores.append(mean_squared_error(y_val_fold, y_pred_val))

    # Return the average MSE across all folds
    return np.mean(mse_scores)

# Create an Optuna study and optimize the objective function
print("Starting Optuna hyperparameter optimization...")
study = optuna.create_study(direction='minimize') # We want to minimize MSE
study.optimize(objective, n_trials=50) # Run 50 trials

print("\n--- Optuna Optimization Results ---")
print(f"Best trial: {study.best_trial.value:.4f} (MSE)")
print("Best hyperparameters found:")
for key, value in study.best_params.items():
    print(f"  {key}: {value}")

# Train the final GradientBoostingRegressor with the best hyperparameters from Optuna
best_gbm_optuna = GradientBoostingRegressor(
    **study.best_params,
    random_state=42
)

# Create a pipeline for the optimized model
optimized_gbm_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('model', best_gbm_optuna)
])

# Fit the pipeline on the full training data
optimized_gbm_pipeline.fit(X_train, y_train)

# Make predictions on the test set
y_pred_optuna_gbm = optimized_gbm_pipeline.predict(X_test)

# Evaluate the optimized model
r2_optuna_gbm = r2_score(y_test, y_pred_optuna_gbm)
mse_optuna_gbm = mean_squared_error(y_test, y_pred_optuna_gbm)

print(f"\n--- Optimized GradientBoostingRegressor (Optuna) Performance on Test Set ---")
print(f"R-squared: {r2_optuna_gbm}")
print(f"Mean Squared Error: {mse_optuna_gbm}")

# Add to results for comparison
results["Gradient Boosting Regressor (Optuna Tuned)"] = {"R-squared": r2_optuna_gbm, "Mean Squared Error": mse_optuna_gbm}

# Update and print the best model based on R-squared
best_model_name = max(results, key=lambda k: results[k]['R-squared'])
print(f"\n--- Best Model Overall based on R-squared: {best_model_name} ---")
print(f"R-squared: {results[best_model_name]['R-squared']}")
print(f"Mean Squared Error: {results[best_model_name]['Mean Squared Error']}")

[I 2026-05-23 05:27:32,315] A new study created in memory with name: no-name-57f4ac39-cdc3-42c6-8b71-31bc338f740f


Starting Optuna hyperparameter optimization...


[I 2026-05-23 05:27:37,507] Trial 0 finished with value: 1304622216487.6558 and parameters: {'n_estimators': 800, 'learning_rate': 0.09337481139902197, 'max_depth': 10, 'min_samples_split': 15, 'min_samples_leaf': 8, 'subsample': 0.7}. Best is trial 0 with value: 1304622216487.6558.
[I 2026-05-23 05:27:43,041] Trial 1 finished with value: 1354285281122.669 and parameters: {'n_estimators': 700, 'learning_rate': 0.14499901571242665, 'max_depth': 8, 'min_samples_split': 18, 'min_samples_leaf': 2, 'subsample': 0.7}. Best is trial 0 with value: 1304622216487.6558.
[I 2026-05-23 05:27:45,890] Trial 2 finished with value: 1131766523183.6335 and parameters: {'n_estimators': 400, 'learning_rate': 0.021818404464819318, 'max_depth': 10, 'min_samples_split': 20, 'min_samples_leaf': 6, 'subsample': 0.8}. Best is trial 2 with value: 1131766523183.6335.
[I 2026-05-23 05:27:55,345] Trial 3 finished with value: 1302170243114.7031 and parameters: {'n_estimators': 1000, 'learning_rate': 0.107767832580222


--- Optuna Optimization Results ---
Best trial: 1082039742552.7214 (MSE)
Best hyperparameters found:
  n_estimators: 400
  learning_rate: 0.011339232288295314
  max_depth: 6
  min_samples_split: 11
  min_samples_leaf: 3
  subsample: 0.6

--- Optimized GradientBoostingRegressor (Optuna) Performance on Test Set ---
R-squared: 0.650853930034004
Mean Squared Error: 1764783336883.557

--- Best Model Overall based on R-squared: Gradient Boosting Regressor ---
R-squared: 0.6646855642239725
Mean Squared Error: 1694870370248.4102


### Final Pipeline for Exporting the Best Model

Based on our comprehensive evaluation, the **Gradient Boosting Regressor** (with its default parameters) has demonstrated the highest performance in terms of R-squared. We will now create a final pipeline incorporating the `StandardScaler` and this best model, then export it as a `model.pkl` file. This serialized model can be loaded later for making predictions on new data without needing to retrain it.

In [87]:
import joblib
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import GradientBoostingRegressor

# Instantiate the best model (untuned GradientBoostingRegressor)
# as it had the highest R-squared in our evaluations
best_final_model = GradientBoostingRegressor(n_estimators=400,learning_rate=0.01,max_depth=6,min_samples_leaf=3,min_samples_split=11,subsample=0.6)

# Create the final pipeline including the scaler and the best model
final_export_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('model', best_final_model)
])

# Fit the final pipeline on the entire training dataset
print("Training the final pipeline...")
final_export_pipeline.fit(X_train, y_train)
print("Final pipeline trained successfully.")

# Export the trained pipeline to a .pkl file
model_filename = 'model.pkl'
joblib.dump(final_export_pipeline, model_filename)

print(f"Final model pipeline exported as '{model_filename}'")

Training the final pipeline...
Final pipeline trained successfully.
Final model pipeline exported as 'model.pkl'


### Creating a Gradio App to Interact with the Model

Now that we have our `model.pkl` file, let's create a simple web application using Gradio. This app will allow us to input various house features and get a price prediction from our deployed model.

In [ ]:
import gradio as gr
import joblib
import pandas as pd
import numpy as np

# Load the saved model pipeline
loaded_pipeline = joblib.load('model.pkl')

# Define the feature columns as they were used during model training
# This list must match the order and names of features X_train was composed of.
FEATURE_COLUMNS = [
    'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad', 'guestroom',
    'basement', 'hotwaterheating', 'airconditioning', 'parking',
    'prefarea', 'furnishingstatus'
]

# Define the prediction function that will be used by Gradio
def predict_price(
    area,
    bedrooms,
    bathrooms,
    stories,
    mainroad,
    guestroom,
    basement,
    hotwaterheating,
    airconditioning,
    parking,
    prefarea,
    furnishingstatus
):
    # Map 'Yes'/'No' inputs to 1/0 for binary categorical features
    mainroad_val = 1 if mainroad == 'Yes' else 0
    guestroom_val = 1 if guestroom == 'Yes' else 0
    basement_val = 1 if basement == 'Yes' else 0
    hotwaterheating_val = 1 if hotwaterheating == 'Yes' else 0
    airconditioning_val = 1 if airconditioning == 'Yes' else 0
    prefarea_val = 1 if prefarea == 'Yes' else 0

    # Map furnishingstatus to its encoded numerical value
    furnishing_map = {'furnished': 0, 'semi-furnished': 1, 'unfurnished': 2}
    furnishingstatus_val = furnishing_map.get(furnishingstatus, 2) # Default to 'unfurnished' if not found

    # Create a DataFrame from the inputs, ensuring correct column order
    input_data = pd.DataFrame([[area, bedrooms, bathrooms, stories, mainroad_val, guestroom_val,
                                  basement_val, hotwaterheating_val, airconditioning_val, parking,
                                  prefarea_val, furnishingstatus_val]],
                              columns=FEATURE_COLUMNS) # Use the defined FEATURE_COLUMNS

    # Make prediction using the loaded pipeline
    prediction = loaded_pipeline.predict(input_data)[0]

    return f"Predicted Price: {prediction:,.2f} INR"

# Define Gradio input components
inputs = [
    gr.Number(label="Area (sq ft)", value=5900),
    gr.Number(label="Bedrooms", value=4, precision=0),
    gr.Number(label="Bathrooms", value=2, precision=0),
    gr.Number(label="Stories", value=2, precision=0),
    gr.Radio(choices=["No", "Yes"], label="Main Road Access", value="No"),
    gr.Radio(choices=["No", "Yes"], label="Guest Room", value="No"),
    gr.Radio(choices=["No", "Yes"], label="Basement", value="Yes"),
    gr.Radio(choices=["No", "Yes"], label="Hot Water Heating", value="No"),
    gr.Radio(choices=["No", "Yes"], label="Air Conditioning", value="No"),
    gr.Number(label="Parking Spaces", value=1, precision=0),
    gr.Radio(choices=["No", "Yes"], label="Preferred Area", value="No"),
    gr.Dropdown(choices=["furnished", "semi-furnished", "unfurnished"], label="Furnishing Status", value="semi-furnished")
]

# Create the Gradio interface
interface = gr.Interface(
    fn=predict_price,
    inputs=inputs,
    outputs="text",
    title="House Price Prediction",
    description="Enter the features of the house to get a price prediction."
)

# Launch the app
if __name__ == "__main__":
    interface.launch(share=True)

### Verifying the Exported Model (`model.pkl`)

Now, let's load the saved `model.pkl` file and use it to make predictions on a small sample of our test data. This will confirm that the model was correctly exported and can be reloaded for inference.

In [88]:
import joblib
import pandas as pd

# Load the saved model pipeline
loaded_pipeline = joblib.load('model.pkl')
print("Model pipeline loaded successfully from 'model.pkl'.")

# Prepare a sample for prediction (using the same sample as before)
new_data_sample_for_check = X_test.head(5)

print("\nSample data for prediction:")
display(new_data_sample_for_check)

# Make predictions using the loaded pipeline
loaded_predictions = loaded_pipeline.predict(new_data_sample_for_check)

print("\nPredictions from the loaded model:")
for i, pred in enumerate(loaded_predictions):
    print(f"Sample {i+1}: Predicted Price = {pred:,.2f}")

# Optionally, compare with actual prices
actual_prices_sample_check = y_test.head(5)
print("\nActual prices for the sample:")
display(actual_prices_sample_check)


Model pipeline loaded successfully from 'model.pkl'.

Sample data for prediction:


,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
316,5900,4,2,2,0,0,1,0,0,1,0,2
77,6500,3,2,3,1,0,0,0,1,0,1,0
360,4040,2,1,1,1,0,0,0,0,0,0,1
90,5000,3,1,2,1,0,0,0,1,0,0,1
493,3960,3,1,1,1,0,0,0,0,0,0,0



Predictions from the loaded model:
Sample 1: Predicted Price = 5,041,155.28
Sample 2: Predicted Price = 7,306,522.60
Sample 3: Predicted Price = 3,698,825.79
Sample 4: Predicted Price = 4,400,590.43
Sample 5: Predicted Price = 4,028,918.61

Actual prices for the sample:


,price
316,4060000
77,6650000
360,3710000
90,6440000
493,2800000
